### ISSUE TITLE
[HMA] Unexplained serialization errors in aws_dataclass.py

### ISSUE THREAD

**Issue by dmukhg** [facebook/ThreatExchange#697]
Filing this issue to keep track of what's going wrong and how I have attempted to address it.

# tl;dr

Long story short. I still don't have a repro that is self-contained. It depends on the import order and what is loaded. It was getting into n^2 number of checks to figure out what's going wrong, so I timed out.

This can be fixed by using `==` instead of `is` to do type checks.
---

When a new dataclass with a field of type [`t.List[SomeNewClass]`](https://github.com/facebook/ThreatExchange/pull/693/commits/33b670db85796419aa5ce207064e5c99945c7aec#diff-c45aea533170253d14a7b23733215edd0286d97b38f4596d0921eb02cdaffcffR231) was added to message_models in hmalib.commons, random tests in aws_dataclasses_test.py started to fail.

The failure would only happen if the t.List subscripted to a user-defined class. Use `t.List[int]` and the issue would fix.

@schatten spent a _LOT_ of time debugging this. A regular debugger would not work because this happened only when all tests were run together. 

Something in the import path of the tests was messing things up.

He narrowed it down to unexplained behavior in [cpython/typing.py](https://github.com/python/cpython/blob/48e3a1d95aee013974121fcafe19816c0e9a41da/Lib/typing.py#L1074-L1080)

Because there is some kind of caching going on, the python `is` check [used in type equality comparisons during py_to_aws](https://github.com/facebook/ThreatExchange/blob/master/hasher-matcher-actioner/hmalib/common/aws_dataclass.py#L87) would return False. 

This is because `is` checks for **identity** which depends on the memory location. It will give the 'wrong' answer if we are somehow able to get two different instances when we call `t.Set[float]`. Internally, the subscript of `t.Set` is a regular `__getitem__` call. This is further made weirder because cpython/typing.py uses a typed lru_cache. Implies `t.Set[ComplexClass] is not t.Set['ComplexClass']`.

This can be fixed by using `==` instead of `is` to do type checks in aws_dataclass.py.

**Comment by Dcallies** [comment]
Woah, this is crazy! This was bothering me all last night trying to think of what could have possibly failed. Excellent debugging work!

**Comment by Dcallies** [comment]
D'oh, this appears to probably be correct behavior (that == is the expected thing), but boils to a clash of conventions. `is` is a very narrow operator, and cannot be messed with by the runtime - the implementation always means "it must be the exact same object".

Python.typing is backed mostly by objects List[x] and List[x] are two declarations of List. It's probably not expected that they be the same object (though due to some caching magic, it appears they often are).

However, conventions for type checking of type is to use `is`. `type(x) is bool` for example. However, the standard to check objects is isinstance(). Python.typing falls somewhere in the middle, and `is` is definitely the wrong one to use :/
